pkg/encryption/path_new.go

```diff
@@ -0,0 +1,290 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package encryption
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/base64"
+	"strings"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/paths"
+	"storj.io/storj/pkg/storj"
+)
+
+// StoreEncryptPath encrypts the path using the provided cipher and looking up
+// keys from the provided store and bucket.
+func StoreEncryptPath(bucket string, path paths.Unencrypted, cipher storj.Cipher, store *Store) (
+	encPath paths.Encrypted, err error) {
+
+	// Invalid paths map to invalid paths
+	if !path.Valid() {
+		return paths.Encrypted{}, nil
+	}
+
+	if cipher == storj.Unencrypted {
+		return paths.NewEncrypted(path.Raw()), nil
+	}
+
+	_, consumed, base := store.LookupUnencrypted(bucket, path)
+	if base == nil {
+		return paths.Encrypted{}, errs.New("unable to find encryption base for: %s/%q", bucket, path)
+	}
+
+	remaining, ok := path.Consume(consumed)
+	if !ok {
+		return paths.Encrypted{}, errs.New("unable to encrypt bucket path: %s/%q", bucket, path)
+	}
+
+	// if we didn't consume any path, we're at the root of the bucket, and so we have
+	// to fold the bucket name into the key.
+	key := &base.Key
+	if !consumed.Valid() {
+		key, err = DeriveKey(key, "path:"+bucket)
```

Comment:
nit: maybe the "path:"+bucket thing should be a function just to prevent typos since it is done in a few spots?